File: setup/kuroshiro.ts
// tslint:disable:no-console

import * as shell from 'shelljs'
import * as fs from 'fs-extra'
import * as path from 'path'

function exec(cmd) {
  if (shell.exec(cmd).code !== 0) throw new Error(`failed: ${cmd}`)
}

const root = path.normalize(path.dirname(path.dirname(__filename)))
const dicts = `${root}/node_modules/kuromoji/dict`
const unzipped = `${root}/build/resource/kuromoji`

exec(`mkdir -p ${unzipped}`)
console.log('copying kuromoji dicts...')
for (const dict of fs.readdirSync(dicts)) {
  console.log('  ', dict)
  const dat = dict.replace('.gz', '')
  exec(`gunzip -c ${dicts}/${dict} > ${unzipped}/${dat}`)
}

console.log('patching kuroshiro')
exec('patch -u -p0 -o node_modules/kuroshiro/src/core-sync.js < setup/kuroshiro.patch')
exec('patch -u -p0 -o node_modules/kuroshiro-analyzer-kuromoji/src/kuroshiro-analyzer-kuromoji-sync.js < setup/kuroshiro-analyzer-kuromoji-sync.patch')

File: setup/kuroshiro.py
#!/usr/bin/env python3

import subprocess
import glob
import os
import shutil
import gzip
import json

root = os.path.join(os.path.dirname(__file__), '..')

dicts_tgt = os.path.join(root, 'build/content/resource/kuromoji')
os.makedirs(dicts_tgt, exist_ok=True)

def run(cmd):
  print(' ', subprocess.check_output(cmd, shell=True).decode('utf-8').strip())

with open(os.path.join(root, 'node_modules/kuromoji/package.json')) as f:
  pkg = json.load(f)
print('copying kuromoji dicts', pkg['version'])

for src in sorted(glob.glob(os.path.join(root, 'node_modules/kuromoji/dict/*.gz'))):
  dict_name = os.path.basename(src)
  print(' ', dict_name)
  tgt = os.path.join(dicts_tgt, dict_name)
  os.makedirs(os.path.dirname(tgt), exist_ok=True)
  #with gzip.open(src, 'rb') as f_in, open(tgt.replace('.gz', ''), 'wb') as f_out:
  #  shutil.copyfileobj(f_in, f_out)
  shutil.copy(src, tgt)

jieba = os.path.join(root, 'build/content/resource/jieba')
os.makedirs(jieba, exist_ok=True)
shutil.copyfile('node_modules/wasmjieba-web/wasmjieba-web_bg.wasm', 'build/content/resource/jieba/wasmjieba-web_bg.wasm')
